File: internal/report/codequality.go
package report

import (
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
)

// GitLab Code Quality output — the CodeClimate-derived JSON that GitLab's
// merge request widget renders natively. It is GitLab's counterpart to
// uploading SARIF on GitHub: findings appear in the MR UI itself, with
// degradation arrows against the target branch, no comment-posting token
// required.
//
// https://docs.gitlab.com/ci/testing/code_quality/#code-quality-report-format

type codeQualityIssue struct {
	Description string              `json:"description"`
	CheckName   string              `json:"check_name"`
	Fingerprint string              `json:"fingerprint"`
	Severity    string              `json:"severity"`
	Location    codeQualityLocation `json:"location"`
}

type codeQualityLocation struct {
	Path  string           `json:"path"`
	Lines codeQualityLines `json:"lines"`
}

type codeQualityLines struct {
	Begin int `json:"begin"`
}

// severityToCodeQuality maps onto GitLab's accepted set
// (info/minor/major/critical/blocker). "blocker" is reserved: this tool's
// notion of blocking lives in its exit code and threshold, and claiming the
// word in a UI that didn't consult the threshold would misstate the tool.
var severityToCodeQuality = map[Severity]string{
	SeverityLow:      "info",
	SeverityMedium:   "minor",
	SeverityHigh:     "major",
	SeverityCritical: "critical",
}

// RenderCodeQuality serialises findings as a GitLab Code Quality report.
//
// The fingerprint deliberately excludes the line number — GitLab diffs
// issues across pipelines by fingerprint, and one that shifted with the
// line would make every rebase look like findings appearing and vanishing.
// It does include the message, unlike the baseline's coarser key: two
// same-category findings on one resource (two hardcoded credentials, say)
// must not collapse into a single widget row.
func RenderCodeQuality(findings []Finding) ([]byte, error) {
	issues := make([]codeQualityIssue, 0, len(findings))
	for _, f := range findings {
		if f.Waived {
			continue // accepted findings are decisions, not open issues
		}
		sum := sha256.Sum256([]byte(string(f.Category) + "\x00" + f.Resource + "\x00" + f.File + "\x00" + f.Message))
		issues = append(issues, codeQualityIssue{
			Description: f.Resource + ": " + f.Message,
			CheckName:   string(f.Category),
			Fingerprint: hex.EncodeToString(sum[:]),
			Severity:    severityToCodeQuality[f.Severity],
			Location: codeQualityLocation{
				Path:  f.File,
				Lines: codeQualityLines{Begin: f.Line},
			},
		})
	}
	return json.MarshalIndent(issues, "", "  ")
}

File: src/tfpdf/report/codequality.py
"""Sortie Code Quality de GitLab.

Port de internal/report/codequality.go.

Le JSON dérivé de CodeClimate que le widget de merge request de GitLab rend
nativement. C'est l'équivalent GitLab de l'envoi de SARIF sur GitHub : les
découvertes apparaissent dans l'interface de la MR elle-même, avec des flèches
de dégradation par rapport à la branche cible, sans qu'aucun jeton de
publication de commentaire ne soit requis.

https://docs.gitlab.com/ci/testing/code_quality/#code-quality-report-format
"""

from __future__ import annotations

import hashlib
from typing import Any

from ._json import marshal_indent
from .finding import Finding, Severity

#: Maps onto GitLab's accepted set (info/minor/major/critical/blocker).
#: "blocker" is reserved: this tool's notion of blocking lives in its exit code
#: and threshold, and claiming the word in a UI that did not consult the
#: threshold would misstate the tool.
SEVERITY_TO_CODE_QUALITY = {
    Severity.LOW: "info",
    Severity.MEDIUM: "minor",
    Severity.HIGH: "major",
    Severity.CRITICAL: "critical",
}


def render_code_quality(findings: list[Finding]) -> bytes:
    """Sérialise les découvertes en un rapport Code Quality GitLab.

    L'empreinte exclut délibérément le numéro de ligne : GitLab compare les
    problèmes d'un pipeline à l'autre par empreinte, et une empreinte qui se
    décalerait avec la ligne ferait ressembler chaque rebase à des découvertes
    qui apparaissent et disparaissent. Elle inclut en revanche le message,
    contrairement à la clé plus grossière de la référence : deux découvertes de
    même catégorie sur une même ressource — deux identifiants en dur, par
    exemple — ne doivent pas se fondre en une seule ligne du widget.
    """
    issues: list[dict[str, Any]] = []
    for finding in findings:
        if finding.waived:
            continue  # accepted findings are decisions, not open issues
        key = "\x00".join([str(finding.category), finding.resource, finding.file, finding.message])
        issues.append(
            {
                "description": finding.resource + ": " + finding.message,
                "check_name": str(finding.category),
                "fingerprint": hashlib.sha256(key.encode()).hexdigest(),
                "severity": SEVERITY_TO_CODE_QUALITY.get(finding.severity, ""),
                "location": {"path": finding.file, "lines": {"begin": finding.line}},
            }
        )
    return marshal_indent(issues)
